New profiling

Duhemm opened the pull request:
This is still work in progress.

Currently, the profile becomes quickly very large (I killed the benchmarks when the memory consumption approached 8GB), but at least it contains a lot of information. Should we log less?

I implemented the parser. Should I split this PR in profiling only and open a second PR with the parser?

densh commented on tools/src/main/scala/scala/scalanative/profiling/Events.scala:
```diff
@@ -0,0 +1,16 @@
+object Event {
+  case class Call(ptr: Long, argc: Int)            extends Event
+  case class Load(ptr: Long)                       extends Event
+  case class Store(ptr: Long)                      extends Event
```
We should try to avoid unstable information like raw pointers in Load, Store, Call, As, Is, Unbox. Those are not stable from run to run and it's going to be quite hard to reason about them.

densh commented on tools/src/main/scala/scala/scalanative/profiling/Events.scala:
```diff
@@ -0,0 +1,16 @@
+
+sealed trait Event
+object Event {
+  case class Call(ptr: Long, argc: Int)            extends Event
```
It might be a better idea to inject those in the beginning of the first basic block, and persist method id instead of the pointer.

densh commented on nativelib/src/main/resources/profiling.c:
```diff
@@ -0,0 +1,220 @@
+    bytes[7] =  value        & 0xFF;
+
+    block_push_many(bytes, 8);
+}
```
We should probably use variable-size encoding like https://source.android.com/devices/tech/dalvik/dex-format.html for ints.

densh commented on tools/src/main/scala/scala/scalanative/profiling/Events.scala:
```diff
@@ -0,0 +1,16 @@
+  case class Load(ptr: Long)                       extends Event
+  case class Store(ptr: Long)                      extends Event
+  case class Classalloc(typeid: Int, name: String) extends Event
+  case class Method(actualTypeid: Int, scopeTypeid: Int, name: String)
```
Lets no persist names, those are big and can be recovered from type/method ids.

densh commented:
Needs a rebase & reformat due to scalafmt upgrade.

densh commented:
Lets delay this one for now. Feel free to reopen once polished.

Duhemm commented:
I can't reopen on my own. I'm not sure if you prefer a new PR or if you can reopen this one, so that the previous review is not lost.

I addressed your comments, but the profile still becomes very large. Here are the profile length for 1 iteration of our benchmarks:

```
bounce   -> 1.6MB
CD       -> 2.2GB
havlak   -> killed when approaching 10GB
json     -> 24MB
list     -> 1.7MB
mandel   -> 329MB
nbody    -> 1.3GB
permute  -> 2.9MB
richards -> 3.9MB
sieve    -> 2.7MB
storage  -> 2.4MB
towers   -> 3.4MB
tracer   -> 11MB
```

The profile for `tracer` contains 2,836,386 events and is parsed in 318,247,964 (~ 112.2 ns/event)

What would be the solution? Adding a new setting to filter the events that should be included?

densh commented:
Hi @Duhemm, 

Do you still keep the whole profile in memory and dump it in the very end of execution? Have you considered writing to disk in batches? 

Feel free to re-open PR and we can continue the discussion over there.
